src/psyclone/configuration.py

```diff
@@ -757,59 +757,47 @@ def __init__(self, config, section):
-                        "[dynamo0.3] section of the config file: {0}".
-                        format(str(err)), config=self._config)
-            elif key == "default_kind":
-                # Set default kinds (precisions) from config file
-                all_kinds = self.create_dict_from_string(section[key])
-                from psyclone.dynamo0p3 import SUPPORTED_FORTRAN_DATATYPES as \
-                    dynamo0p3_datatypes
-                # Check for valid datatypes (filter to remove empty values)
-                datatypes = list(filter(None, all_kinds.keys()))
-                if datatypes != dynamo0p3_datatypes:
-                    raise ConfigurationError(
-                        "Invalid datatype found in the '[dynamo0.3]' "
-                        "section of the configuration file '{0}'. "
-                        "Valid datatypes are: '{1}'."
-                        .format(config.filename, dynamo0p3_datatypes))
-                # Check for valid kinds (filter to remove any empty values)
-                datakinds = list(filter(None, all_kinds.values()))
-                if len(datakinds) != len(dynamo0p3_datatypes):
-                    raise ConfigurationError(
-                        "Did not find default kind for one or more "
-                        "datatypes in the '[dynamo0.3]' section "
-                        "of the configuration file '{0}'."
-                        .format(config.filename))
-                self._default_kind = all_kinds
-            else:
+        # Parse setting for redundant computation over annexed dofs
+        try:
+            self._compute_annexed_dofs = section.getboolean(
+                'COMPUTE_ANNEXED_DOFS')
+        except ValueError as err:
+            raise ConfigurationError(
+                "error while parsing COMPUTE_ANNEXED_DOFS in the [dynamo0.3] "
+                "section of the config file: {0}".format(str(err)),
+                config=self._config)
+
+        # Parse setting for default kinds (precisions)
+        if "default_kind" in section:
+            # Set default kinds (precisions) from config file
+            all_kinds = self.create_dict_from_string(section["default_kind"])
+            from psyclone.dynamo0p3 import SUPPORTED_FORTRAN_DATATYPES
+            # Check for valid datatypes (filter to remove empty values)
+            datatypes = list(filter(None, all_kinds.keys()))
+            if datatypes != SUPPORTED_FORTRAN_DATATYPES:
+                raise ConfigurationError(
+                    "Invalid datatype found in the '[dynamo0.3]' "
+                    "section of the configuration file '{0}'. "
+                    "Valid datatypes are: '{1}'."
+                    .format(config.filename, SUPPORTED_FORTRAN_DATATYPES))
+            # Check for valid kinds (filter to remove any empty values)
+            datakinds = list(filter(None, all_kinds.values()))
+            if len(datakinds) != len(SUPPORTED_FORTRAN_DATATYPES):
```

Comment:
Similarly, if no 'default_kind' is provided in the input file then the code falls over because it assumes there are always values in the Config object. I think we should mandate that default_kind always be provided and therefore alter this bit of code slightly so as to check. Please then also extend the testing so that we check that `default_kind` is mandatory.

Reply:
Also good point but it opened a can of worms.  Besides the `default_kind`, mandatory settings for LFRic are also `COMPUTE_ANNEXED_DOFS` and `access_mapping`. I reformulated the code to check for presence of all mandatory settings, which then required modifications in `config_test.py` and `dummy_config.cfg`. Since `DynConfig` accepts both `"dynamo0p1"` and `"dynamo0p3"` API I enclosed parsing of annexed dofs and default kind into an `if` clause. I did not bother with checks for Dynamo0.1 API since it is going to be removed in #282, but I left "TODO" comments for removal if the `if` clause.
Finally, having an invalid or incomplete key/value mapping format (e.g. `DEFAULTAPI` instead of `DEFAULTAPI = dynamo0.3`) in config file triggers `configparser` general `ParsingError` which was not caught previously and is not covered by neither the missing mandatory setting nor the `MissingSectionHeaderError` check. I added this exception and extended the `test_broken_fmt` in `config_test.py` to check for it.